### PR TITLE
net: lwm2m: refactoring lwm2m_engine.c

### PR DESCRIPTION
**Pull request by otkulseng**
Created three files and split lwm2m_engine.c into the four files:
- lwm2m_registry.c, creation/deletion and setters/getters.
- lwm2m_message_handling.c, CoAP and message handling.
- lwm2m_observation.c, all code regarding observations.
- lwm2m_engine.c, socket (including lwm2m main work loop).

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

### CODE REVIEW

**Comment by otkulseng** [comment]
The checkpatch failures above are legacy errors. What should I do with them?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 1]
```diff
@@ -232,5024 +169,464 @@ static char *sprint_token(const uint8_t *token, uint8_t tkl)
...
+		/* Bootstrap server have a access to Security and Server object */
+		return true;
 	}
 
 	return false;
 }
 
-int lwm2m_notify_observer(uint16_t obj_id, uint16_t obj_inst_id, uint16_t res_id)
+/* This function is exposed for the content format writers */
+int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst,
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:303]: Shouldn't this be moved along with `lwm2m_read_handler()`?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -232,5024 +169,464 @@ static char *sprint_token(const uint8_t *token, uint8_t tkl)
...
-		if (coap_header_get_type(&response) == COAP_TYPE_CON) {
-			r = lwm2m_send_empty_ack(client_ctx,
-						 coap_header_get_id(&response));
-			if (r < 0) {
-				LOG_ERR("Error transmitting ACK");
-			}
-		}
+	return true;
+}
+static int bootstrap_delete(struct lwm2m_message *msg)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:582]: It looks it won't build with bootstrap, this is static, yet called from within `lwm2m_message_handling.c`

File: subsys/net/lib/lwm2m/lwm2m_registry.c [hunk 3]
```diff
@@ -0,0 +1,1321 @@
...
+		if (obj_inst->obj->obj_id == obj_id &&
+		    obj_inst->obj_inst_id == obj_inst_id) {
+			return obj_inst;
+		}
+	}
+
+	return NULL;
+}
+struct lwm2m_engine_obj_inst *
+next_engine_obj_inst(int obj_id, int obj_inst_id)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_registry.c:136]: This line break seems weird - perhaps given the occasion try running `git-clang-format` to let it do the work?

**Review by rlubos — COMMENTED** [review]
A very brief initial look, it seems we should/could slim `lwm2m_engine.c` even further?

Looking at the changes I also wonder if we should specify some kind of `lwm2m_transport.c` module to deal with the sockets? This is up for consideration though, no push.

**Comment by rlubos** [comment]
Another consideration for further cleanup - maybe we could specify some subdirectories in the `lwm2m` folder so that we don't dump everyting together. I was thinking of a separte subdir for `engine` stuff (reworked in this PR), separate for `objects` implementations and separate for `encoders`

**Comment by otkulseng** [comment]
> Another consideration for further cleanup - maybe we could specify some subdirectories in the `lwm2m` folder so that we don't dump everyting together. I was thinking of a separte subdir for `engine` stuff (reworked in this PR), separate for `objects` implementations and separate for `encoders`

Do you maybe think it wise to do this in a future pr, so we can limit the size of this one?

**Review by SeppoTakalo — DISMISSED** [review]

File: subsys/net/lib/lwm2m/lwm2m_observation.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+};
+/* Attribute handling. */
+
+struct lwm2m_attr *lwm2m_engine_get_next_attr(const void *ref,
+					      struct lwm2m_attr *prev);
+const char *lwm2m_engine_get_attr_name(const struct lwm2m_attr *attr);
+
+void clear_attrs(void *ref);
+
+/**/
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_observation.h:36]: nit: Empty comment?

File: subsys/net/lib/lwm2m/lwm2m_observation.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+int64_t engine_observe_shedule_next_event(struct observe_node *obs, uint16_t srv_obj_inst,
+						 const int64_t timestamp);
+
+void remove_observer_from_list(struct lwm2m_ctx *ctx, sys_snode_t *prev_node,
+				      struct observe_node *obs);
+
+struct observe_node *engine_observe_node_discover(sys_slist_t *observe_node_list,
+							sys_snode_t **prev_node,
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_observation.h:45]: Would be nice to run this file over `clang-format` to fix the ailgnment issues like this.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+struct lwm2m_message *find_msg(struct coap_pending *pending,
+				      struct coap_reply *reply);
+void lwm2m_reset_message(struct lwm2m_message *msg, bool release);
+void lm2m_message_clear_allocations(struct lwm2m_message *msg);
+int lwm2m_init_message(struct lwm2m_message *msg);
+int lwm2m_send_message(struct lwm2m_message *msg);
+int lwm2m_send_message_async(struct lwm2m_message *msg);
+/* */
+int handle_request(struct coap_packet *request,
+			  struct lwm2m_message *msg);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.h:50]: ditto, would be good to run it over `clang-format`

**Review by rlubos — DISMISSED** [review]
Thanks, spotted some nits but nothing serious.

**Comment by otkulseng** [comment]
Included lwm2m_observation.h in lwm2m_message_handling.h as the function int generate_notify_message(struct lwm2m_ctx *ctx, struct observe_node *obs, void *user_data); uses the observe_node struct definined in _observation.h

**Review by SeppoTakalo — APPROVED** [review]

**Review by VeijoPesonen — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**